File: frontend/src/api/EventService.ts
import type { CreateEventRequest, EventResponse, UpdateEventRequest, RsvpRequest } from '../types/events';
import type { StatusResponse, PaginatedResponse } from '../types/common';
import { apiClient } from './apiClient';

export class EventService {
  static async createEvent(request: CreateEventRequest): Promise<EventResponse> {
    return apiClient<EventResponse>('/api/events', {
      method: 'POST',
      body: JSON.stringify(request),
    });
  }

  static async getAllEvents(limit?: number, offset?: number, organizationId?: number, startDate?: String, endDate?: String, sortOrder?: string): Promise<PaginatedResponse<EventResponse>> {
    const url = new URL('/api/events', window.location.origin);
    if (limit) {
      url.searchParams.append('limit', limit.toString());
    }
    if (offset) {
      url.searchParams.append('offset', offset.toString());
    }
    if (organizationId) {
      url.searchParams.append('organizationId', organizationId.toString());
    }
    if (startDate) {
      url.searchParams.append('startDate', startDate.toString());
    }
    if (endDate) {
      url.searchParams.append('endDate', endDate.toString());
    }
    if (sortOrder) {
      url.searchParams.append('sortOrder', sortOrder);
    }
    return apiClient<PaginatedResponse<EventResponse>>(url.toString());
  }

  static async getEvent(eventId: number): Promise<EventResponse> {
    return apiClient<EventResponse>(`/api/events/${eventId}`);
  }

  static async updateEvent(eventId: number, request: UpdateEventRequest): Promise<EventResponse> {
    return apiClient<EventResponse>(`/api/events/${eventId}`, {
      method: 'PUT',
      body: JSON.stringify(request),
    });
  }

  static async deleteEvent(eventId: number): Promise<StatusResponse> {
    return apiClient<StatusResponse>(`/api/events/${eventId}`, { method: 'DELETE' });
  }

  static async rsvpToEvent(eventId: number, request: RsvpRequest): Promise<StatusResponse> {
    return apiClient<StatusResponse>(`/api/events/${eventId}/rsvp`, {
      method: 'POST',
      body: JSON.stringify(request),
    });
  }

  static async joinEvent(eventId: number): Promise<StatusResponse> {
    return apiClient<StatusResponse>(`/api/events/${eventId}/join`, { method: 'POST' });
  }

  static async getRequestedRsvps(): Promise<EventResponse[]> {
    return apiClient<EventResponse[]>('/api/events/rsvps/requested');
  }
}

File: backend/service/src/main/kotlin/alphainterplanetary/teven/service/event/EventService.kt
package alphainterplanetary.teven.service.event

import alphainterplanetary.teven.api.model.event.CreateEventRequest
import alphainterplanetary.teven.api.model.event.EventResponse
import alphainterplanetary.teven.api.model.event.UpdateEventRequest
import alphainterplanetary.teven.core.security.AuthContext
import alphainterplanetary.teven.core.security.Permission
import alphainterplanetary.teven.data.event.EventDao

class EventService(private val eventDao: EventDao) {
  suspend fun createEvent(
    authContext: AuthContext,
    createEventRequest: CreateEventRequest,
  ): EventResponse {
    if (createEventRequest.title.isBlank()) {
      throw IllegalArgumentException("Event title cannot be empty")
    }

    val organizationIdToUse = if (authContext.hasPermission(Permission.MANAGE_EVENTS_GLOBAL)) {
      createEventRequest.organizationId
        ?: throw IllegalArgumentException("organizationId is required for MANAGE_EVENTS_GLOBAL users")
    } else {
      authContext.organizationId
    }

    // Create a new request with the determined organizationId
    val requestWithOrgId = createEventRequest.copy(organizationId = organizationIdToUse)

    return eventDao.createEvent(requestWithOrgId)
  }

  suspend fun getEvents(
    authContext: AuthContext,
    startDate: String?,
    endDate: String?,
    limit: Int?,
    offset: Long?,
    sortOrder: String?,
  ): List<EventResponse> {
    val organizationId = if (authContext.hasPermission(Permission.VIEW_EVENTS_GLOBAL)) {
      null
    } else {
      authContext.organizationId
    }
    return eventDao.getEvents(
      organizationId = organizationId,
      startDate = startDate,
      endDate = endDate,
      limit = limit,
      offset = offset,
      sortOrder = sortOrder,
    )
  }

  suspend fun getEventById(authContext: AuthContext, eventId: Int): EventResponse? {
    val event = eventDao.getEventById(eventId)

    return if (event != null && !authContext.hasPermission(Permission.VIEW_EVENTS_GLOBAL) && event.organization.organizationId != authContext.organizationId) {
      null // Not found or not authorized
    } else {
      event
    }
  }

  suspend fun updateEvent(
    authContext: AuthContext,
    eventId: Int,
    updateEventRequest: UpdateEventRequest,
  ): Boolean {
    // Check if the event belongs to the user's organization or if the user has MANAGE_EVENTS_GLOBAL
    val event = eventDao.getEventById(eventId)
      ?: return false // Event not found

    if (!authContext.hasPermission(Permission.MANAGE_EVENTS_GLOBAL) && event.organization.organizationId != authContext.organizationId) {
      throw IllegalAccessException("User not authorized to update events outside their organization")
    }

    // If MANAGE_EVENTS_GLOBAL, allow organizationId to be updated, otherwise ensure it's not changed
    val requestWithOrgId = if (authContext.hasPermission(Permission.MANAGE_EVENTS_GLOBAL)) {
      updateEventRequest
    } else {
      if (updateEventRequest.organizationId != null && updateEventRequest.organizationId != event.organization.organizationId) {
        throw IllegalArgumentException("Non-MANAGE_EVENTS_GLOBAL users cannot change event organizationId")
      }
      updateEventRequest.copy(organizationId = event.organization.organizationId)
    }

    return eventDao.updateEvent(eventId, requestWithOrgId)
  }

  suspend fun deleteEvent(authContext: AuthContext, eventId: Int): Boolean {
    val event = eventDao.getEventById(eventId)
      ?: return false // Event not found

    if (!authContext.hasPermission(Permission.MANAGE_EVENTS_GLOBAL) && event.organization.organizationId != authContext.organizationId) {
      throw IllegalAccessException("User not authorized to delete events outside their organization")
    }
    return eventDao.deleteEvent(eventId)
  }

  suspend fun assignStaffToEvent(authContext: AuthContext, eventId: Int, userId: Int): Boolean {
    val event = eventDao.getEventById(eventId)
      ?: throw IllegalArgumentException("Event not found")

    if (!authContext.hasPermission(Permission.MANAGE_EVENTS_ORGANIZATION) && event.organization.organizationId != authContext.organizationId) {
      throw IllegalAccessException("User not authorized to assign staff to events outside their organization")
    }
    return eventDao.assignStaffToEvent(eventId, userId)
  }

  suspend fun removeStaffFromEvent(authContext: AuthContext, eventId: Int, userId: Int): Boolean {
    val event = eventDao.getEventById(eventId)
      ?: throw IllegalArgumentException("Event not found")

    if (!authContext.hasPermission(Permission.MANAGE_EVENTS_ORGANIZATION) && event.organization.organizationId != authContext.organizationId) {
      throw IllegalAccessException("User not authorized to remove staff from events outside their organization")
    }
    return eventDao.removeStaffFromEvent(eventId, userId)
  }

  suspend fun rsvpToEvent(
    authContext: AuthContext,
    eventId: Int,
    userId: Int,
    availability: String,
  ): Boolean {
    val event = eventDao.getEventById(eventId)
      ?: throw IllegalArgumentException("Event not found")

    if (!authContext.hasPermission(Permission.MANAGE_EVENTS_ORGANIZATION) && event.organization.organizationId != authContext.organizationId) {
      throw IllegalAccessException("User not authorized to RSVP to events outside their organization")
    }
    return eventDao.rsvpToEvent(eventId, userId, availability)
  }
}
